### ISSUE TITLE
Applying plugin suggestions makes build slower to evaluate

### ISSUE THREAD

**Issue by plastiv** [autonomousapps/dependency-analysis-android-gradle-plugin#383]
I don't think this is an issue with plugin so it is not a bug report. I assume people here are a community interested in topic/are build performance experts so I would enjoy to have a discussion around what could have been an underlying issue causing such regression. Let me know if github issues are not an appropriate medium for this. Maybe you have slack/channel for such a discussions somewhere?

I am experimenting with a closed-source 100% kotlin android project with ~115k cloc and ~70 gradle modules. Dagger, dagger-android, databinding, moshi-codegen, parcelize are being used from annotation processors. 

I haven't changed any source code of the project, just applied every api/implementation/unused suggestion. Resulted in 64 files -471 +864 lines changes. 

I'm using gradle profiler 0.16 to compare before/after with next scenario

```
configuration {
  tasks = ["help"]
}

assemble_up_to_date {
  tasks = [":main:assembleDebug"]
}

assemble_clean {
  tasks = [":main:assembleDebug"]
  cleanup-tasks = ["clean"]
}

assemble_rerun_tasks {
  tasks = [":main:assembleDebug"]
  gradle-args = ["--rerun-tasks"]
}

assemble_non_abi_change {
  tasks = [":main:assembleDebug"]
  apply-non-abi-change-to = ["common/ui/src/main/java/package_name/common/ui/view/countrypicker/ui/CountryPickerFragment.kt"]
}

assemble_abi_change {
  tasks = [":main:assembleDebug"]
  apply-abi-change-to = ["common/ui/src/main/java/package_name/common/ui/view/countrypicker/ui/CountryPickerFragment.kt"]
}

assemble_resource_change {
  tasks = [":main:assembleDebug"]
  apply-android-resource-change-to = "common/ui/src/main/res/values/strings.xml"
}

assemble_layout_change {
  tasks = [":main:assembleDebug"]
  apply-android-layout-change-to = "common/ui/src/main/res/layout/fragment_country_picker.xml"
}
```

Scenario | Before | After
------------ | ------------- | -------
configuration |  3,311.73 ms | 5,040.16 ms |
assemble_up_to_date | 6,414.98 ms | 11,088.51 ms |
assemble_clean | 13,736.07 ms |  15,853.22 ms |
assemble_rerun_tasks | 162,105.59 ms |  156,763.41 ms |
assemble_non_abi_change | 13,536.01 ms | 14,091.62 ms |
assemble_abi_change | 19,816.41 ms | 18,490.80 ms |
assemble_resource_change | 10,753.56 ms | 11,194.92 ms |
assemble_layout_change | 87,096.05 ms |  82,542.69 ms |

It looks like "heavy" tasks - abi change, layout change which triggers abi change due to databinding and  rerun tasks are being improved. While "easy" tasks - configuration, up-to-date, local cache, non-abi-change are get slower. Do you think this is expected?

**Comment by autonomousapps** [comment]
This is fascinating, thank you for the data. Which version of the plugin are you using? The most recent, 0.72.0, has a new "mode", that I refer to as non-strict, that seeks to reduce the number of edges in the dependency graph. The plugin's default behavior, ie strict mode, encourages very strict dependency declarations, which can result in duplicate edges in the graph. Consider this simple example:

```
A -> B -> C
```
A depends on B which depends on C. B exposes C to A as an api dependency. I.e.
```groovy
// b/build.gradle
dependencies {
  api project(C)
}
```
So, C is on A's compile classpath. Now, if A actually directly uses C, then strict mode will suggest you add a direct edge from A to C, sort of like this:
```
A -> B -> C
A -> C
```
You can see that the resultant graph has the same number of nodes (3), but one additional edge (3 instead of 2). It's possible this is the source of the performance regression you observed.

The new, non-strict mode, would only suggest adding this new edge if and only if C is part of A's ABI. In other words:
```kotlin
public class A {
  public fun foo(bar: C) // C is a parameter of a public function of A, therefore it's in A's ABI
}
```
In this case, the non-strict mode still suggests to add a direct dependency from A -> C. But in a different case:
```kotlin
public class A {
  private fun foo(bar: C) // C is a parameter of a private function of A, and is therefore not part of the ABI
}
```
in this case, non-strict mode will _not_ suggest adding a direct dependency from A to C.

The other thing 0.72.0 does is it prints some basic summary statistics indicating the size of the graph before and after, should you follow the advice. 

**To be clear, at this point I'm only speculating that strict mode can lead to worse performance in some scenarios**, but the logic makes intuitive sense. What would be really useful at this point would be to have actual CPU and memory profiles of the before and after scenarios, which can also be accomplished with gradle profiler.

**Comment by plastiv** [comment]
Thank you for a detailed reply. It gave me thoughts to process already. I intent to followup on your suggestion to try non-strict mode as well as capture some cpu and memory profiles. 

Currently stopped at trying to figure out this error https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin/issues/386 which was already reported by someone.

**Comment by autonomousapps** [comment]
I can provide advice if you need it for the CPU and memory profiles. The tl;dr is to use [gradle-profiler](https://github.com/gradle/gradle-profiler).

**Comment by plastiv** [comment]
Thank you for offering help!

Let me describe my steps to double check if I'm going into right direction.

Gradle 7, AGP 4.1.3, JDK 11, Kotlin 1.4.32, dependency-analysis-plugin 0.72. 

I'm using snapshot gradle-profiler-0.17.0-alpha04 to get new async-profiler v2 integration which can do CPU and heap flames together. 

I have selected `./gradlew help` as target task in hopes that resulting reports would be easier to investigate. Unfortunately, they don't produce 2 sec difference anymore at the moment. I will rerun full benchmark later to confirm that other tasks difference is still there. 

Before: main
After: main + strict suggestions applied

Benchmark: 2 warm-ups, 5 iterations, `./gradlew help`
Profile: 2 warm-ups, 1 measurement, async-profiler-all, `./gradlew help`

[gradle-profiler-results.zip](https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin/files/6414374/gradle-profiler-results.zip)

**Comment by autonomousapps** [comment]
Those steps sound good! I'll take a look at the zip file later.

**Comment by plastiv** [comment]
I have rerun full benchmark with every scenario and I don't see the difference anymore 😕. 

It looks like I don't control experiment environment to have reproducible results. Pity.

I feel that as I'm learning and stability of the experiment has improved (low standard deviation) but until I will manage to reproduce again the issue can be closed / paused to give room for other discussions.

I would like to thank you for sharing about nodes vs edges and strict vs non-strict mode. TIL

**Comment by plastiv** [comment]
I want to mention that changes I think (can't confirm) have influenced the experiment results are:

* Disabling spotlight indexing (I'm on mac) `System Preferences → Spotlight → Privacy → Exclude folder with projects`.
* Making sure AS is closed during run to avoid IntelliJ re-indexing project.
* Checking temperature to make sure CPU was not throttled during run `watch -n 30 'pmset -g therm | tee -a therm.log'`.
